### PR TITLE
Typescript Rework

### PR DESCRIPTION
**Pull request by reflog**
#### Summary
This PR is a migration of Mattermost-Redux to Typescript

#### Test Information
This PR was tested on: Arch Linux

### CODE REVIEW

File: src/actions/schemes.test.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ import nock from 'nock';
 
 import * as Actions from 'actions/schemes';
 import {Client4} from 'client';
-import {RequestStatus} from 'constants';
+import {RequestStatus} from '../constants';
```
Comment (mickmister) [inline src/actions/schemes.test.js]: Do you know why we have to import `constants` with a relative path now?

File: src/action_types/channels.ts [hunk 2]
```diff
@@ -1,49 +1,35 @@
...
 // See LICENSE.txt for license information.
-// @flow
-
 import keyMirror from 'utils/key_mirror';
-
 export default keyMirror({
     CHANNEL_REQUEST: null,
     CHANNEL_SUCCESS: null,
     CHANNEL_FAILURE: null,
-
```
Comment (mickmister) [inline src/action_types/channels.ts]: The whitespace removed in the action type files kind of bloats the PR, and arguably makes the code harder to read. Do you think the blank lines served a good purpose?

0/5 We leave the newlines in the code

File: src/reducers/entities/roles.ts [hunk 3]
```diff
@@ -1,45 +1,56 @@
...
         return state;
     }
-    case RoleTypes.RECEIVED_ROLE: {
+
+    case RoleTypes.RECEIVED_ROLE:
+    {
         if (action.data) {
-            const nextState = {...state};
+            const nextState = {...state,
+            };
```
Comment (mickmister) [inline src/reducers/entities/roles.ts]: This instance of this notation sort of looks sort of sloppy. I think the application's convention is to have this on one line.

A lot of single-line -> multi-line and multi-line -> single-line changes make this PR bigger than necessary imo. Were a lot of these changes done with an automated tool?

**Comment by mickmister** [comment]
@reflog Do you know if all the npm scripts still work? Namely `npm run dev:watch` and `npm run dev-mobile:watch`

File: src/actions/schemes.test.js (hunk 1 above)
Comment (reflog) [reply src/actions/schemes.test.js]: Yea, this was unfortunate, but 'constants' comes from Typescript itself: node_modules/@types/node/constants.d.ts

File: src/reducers/entities/roles.ts (hunk 3 above)
Comment (reflog) [reply src/reducers/entities/roles.ts]: This is a result of automatic conversion. I've tried to fix most of the formatting issues, but some are still lingering, not sure how to auto-fix this stuff.

File: src/action_types/channels.ts (hunk 2 above)
Comment (reflog) [reply src/action_types/channels.ts]: This is a result of automatic conversion. I've tried to fix most of the formatting issues, but some are still lingering, not sure how to auto-fix this stuff.

**Comment by reflog** [comment]
@jespino - tested webapp and mobile. both work!

File: package.json [hunk 4]
```diff
@@ -8,99 +8,91 @@
   "bugs": {
     "url": "https://github.com/mattermost/mattermost-redux/issues"
   },
+  "main": "build/mattermost.client4.js",
+  "types": "build/index.d.ts",
+  "files": [
+    "build"
```
Comment (anindha) [inline package.json]: When I include this branch in mattermost-webapp some of the expected files are missing at the tests fail. I think it has to do with this line. Do you have a branch of mattermost-webapp that works with this mattermost-redux branch.

Comment (reflog) [reply package.json]: @anindha - can you elaborate on "some of the expected files are missing at the tests fail" ?
If you have some error logs - please paste them here.
This branch should work with regular webapp without changes

Comment (anindha) [reply package.json]: @reflog Here is an example. Could be my setup. Currently pointing webapp master at your branch. This is my package.json:
`"mattermost-redux": "github:reflog/mattermost-redux#typescript",`


`FAIL  components/channel_header_mobile/channel_info_button/channel_info_button.test.js
  ● Test suite failed to run

    Cannot find module 'react-hot-loader' from 'permissions.js'

    However, Jest was able to find:
        './channel_info_button.js'
        './channel_info_button.test.js'

    You might want to include a file extension in your import, or update your 'moduleFileExtensions', which is currently ['ts', 'tsx', 'js', 'jsx', 'json'].

    See https://jestjs.io/docs/en/configuration#modulefileextensions-array-string

      2 | 
      3 | Object.defineProperty(exports, "__esModule", {
    > 4 |   value: true
        |                                  ^
      5 | });
      6 | exports.default = void 0;
      7 | // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.

      at Resolver.resolveModule (node_modules/jest-resolve/build/index.js:230:17)
      at ../mattermost-redux/constants/permissions.js:4:91
      at Object.<anonymous> (../mattermost-redux/constants/permissions.js:6:3)`

Comment (anindha) [reply package.json]: Here is an error in the console when running the `npm run dev-server`. The app fails to run.

```client?9298:159 ./actions/websocket_actions.jsx
Module not found: Error: Can't resolve 'mattermost-redux/action_types' in '/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/actions'```

Comment (reflog) [reply package.json]: @anindha - thanks for the logs! I've fixed the issue, can you try now?

Comment (anindha) [reply package.json]: @reflog Still having the issue

Comment (reflog) [reply package.json]: Can you provide more details please?

Comment (anindha) [reply package.json]: Seems to be a problem how the include path is being resolved. Here is an example.

```
ERROR in ./components/user_settings/security/user_access_token_section/user_access_token_section.jsx
Module not found: Error: Can't resolve 'mattermost-redux/utils/user_utils' in '/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/user_settings/security/user_access_token_section'
resolve 'mattermost-redux/utils/user_utils' in '/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/user_settings/security/user_access_token_section'
  Parsed request is a module
  using description file: /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/package.json (relative path: ./components/user_settings/security/user_access_token_section)
    resolve as module
      /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/user_settings/security/user_access_token_section/node_modules doesn't exist or is not a directory
      /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/user_settings/security/node_modules doesn't exist or is not a directory
      /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/user_settings/node_modules doesn't exist or is not a directory
      /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/node_modules doesn't exist or is not a directory
      /Users/anindha/go/src/github.com/mattermost/node_modules doesn't exist or is not a directory
      /Users/anindha/go/src/github.com/node_modules doesn't exist or is not a directory
      /Users/anindha/go/src/node_modules doesn't exist or is not a directory
      /Users/anindha/go/node_modules doesn't exist or is not a directory
      /Users/anindha/node_modules doesn't exist or is not a directory
      /Users/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory
      looking for modules in /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules
        using description file: /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/package.json (relative path: ./node_modules)
          using description file: /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/package.json (relative path: ./utils/user_utils)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils.js doesn't exist
            .jsx
              Field 'browser' doesn't contain a valid alias configuration
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils.jsx doesn't exist
            .ts
              Field 'browser' doesn't contain a valid alias configuration
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils.ts doesn't exist
            .tsx
              Field 'browser' doesn't contain a valid alias configuration
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils.tsx doesn't exist
            as directory
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils doesn't exist
      looking for modules in /Users/anindha/go/src/github.com/mattermost/mattermost-webapp
        using description file: /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/package.json (relative path: .)
          using description file: /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/package.json (relative path: ./mattermost-redux/utils/user_utils)
            no extension
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils doesn't exist
            .js
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils.js doesn't exist
            .jsx
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils.jsx doesn't exist
            .ts
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils.ts doesn't exist
            .tsx
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils.tsx doesn't exist
            as directory
              /Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils doesn't exist
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/user_settings/security/user_access_token_section/node_modules]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/user_settings/security/node_modules]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/user_settings/node_modules]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/components/node_modules]
[/Users/anindha/go/src/github.com/mattermost/node_modules]
[/Users/anindha/go/src/github.com/node_modules]
[/Users/anindha/go/src/node_modules]
[/Users/anindha/go/node_modules]
[/Users/anindha/node_modules]
[/Users/node_modules]
[/node_modules]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils.js]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils.jsx]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils.ts]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/node_modules/mattermost-redux/utils/user_utils.tsx]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils.js]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils.jsx]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils.ts]
[/Users/anindha/go/src/github.com/mattermost/mattermost-webapp/mattermost-redux/utils/user_utils.tsx]
 @ ./components/user_settings/security/user_access_token_section/user_access_token_section.jsx 11:0-63 127:11-34
 @ ./components/user_settings/security/user_access_token_section/index.js
 @ ./components/user_settings/security/user_settings_security.jsx
 @ ./components/user_settings/security/index.js
 @ ./components/user_settings/user_settings.jsx
 @ ./node_modules/babel-loader/lib??ref--4!./components/user_settings/index.js
 @ ./node_modules/bundle-loader?lazy!./components/user_settings/index.js
 @ ./components/user_settings/modal/user_settings_modal.jsx
 @ ./components/user_settings/modal/index.js
 @ ./components/sidebar/header/dropdown/sidebar_header_dropdown.jsx
 @ ./components/sidebar/header/dropdown/index.js
 @ ./components/sidebar/header/sidebar_header.jsx
 @ ./components/sidebar/header/index.js
 @ ./components/sidebar/sidebar.jsx
 @ ./components/sidebar/index.js
 @ ./components/channel_layout/channel_controller.jsx
 @ ./components/needs_team/needs_team.jsx
 @ ./components/needs_team/index.js
 @ ./components/root/root.jsx
 @ ./node_modules/babel-loader/lib??ref--4!./components/root/index.js
 @ ./node_modules/bundle-loader?lazy!./components/root/index.js
 @ ./components/app.jsx
 @ ./root.jsx
 @ multi ./root.jsx root.html
```

**Comment by reflog** [comment]
@jespino , @mickmister - I've removed most of the formatting issues, so the PR should be much saner now. Would you mind testing it locally (preferrable with webapp or mobile) before we merge?
@anindha - I am unable to reproduce this issue. Are you testing with a clean checkout? If so - can you paste exactly the sequence of commands? (clone, npm install/link, build)

File: src/utils/deep_freeze.ts [hunk 5]
```diff
@@ -8,11 +7,9 @@
  * of patent rights can be found in the PATENTS file in the same directory.
  *
  */
-/* eslint-disable header/header */
 
-const disableOnIe11 =
-    window && window.navigator && window.navigator.userAgent &&
-    window.navigator.userAgent.indexOf('Trident') !== -1;
+/* eslint-disable header/header */
+const disableOnIe11 = window && window.navigator && window.navigator.userAgent && window.navigator.userAgent.indexOf('Trident') !== -1;
```
Comment (devinbinnie) [inline src/utils/deep_freeze.ts]: We can probably just remove this now :)

**Comment by anindha** [comment]
> @anindha - I am unable to reproduce this issue. Are you testing with a clean checkout? If so - can you paste exactly the sequence of commands? (clone, npm install/link, build)

@reflog Do you have a branch of the webapp that I can checkout that points to your redux typescript branch? It might be easier if we are testing on the same branch.

File: src/actions/channels.ts [hunk 6]
```diff
@@ -172,7 +169,7 @@ export function createGroupChannel(userIds: Array<string>): ActionFunc {
             return {error};
         }
 
-        let member = {
+        let member: any = {
```
[the remark this answers is not in the document]
Comment (reflog) [reply src/actions/channels.ts]: 👍

**Comment by zhangriyueming** [comment]
Anyone help me to update: why migrate to typescript from flow?

**Pull request merged**